File: java/340_LongestSubstringWithAtMostKDistinctCharacters.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        if (s.length()*k == 0) return 0;
        
        HashMap<Character, Integer> letCounts = new HashMap<Character, Integer>();
        int kCount = 0, trailingIdx = 0, highCount = 1;
        
        for (int i = 0; i < s.length(); ++i) {
            if (letCounts.get(s.charAt(i)) == null || letCounts.get(s.charAt(i)) == 0) {
                ++kCount;
                letCounts.put(s.charAt(i), 1);
            } else {
                letCounts.put(s.charAt(i), letCounts.get(s.charAt(i))+1);
            }
            
            
            while (trailingIdx < s.length() && kCount > k) {
                char first = s.charAt(trailingIdx);
                while (trailingIdx < s.length() && first == s.charAt(trailingIdx)) {
                    ++trailingIdx;
                    letCounts.put(first, letCounts.get(first)-1);
                }
                    
                if (letCounts.get(first) == 0) --kCount;
            }
            if (highCount < i-trailingIdx+1) highCount = i-trailingIdx+1;
        }
        return highCount;
    }
}

File: cpp/340_LongestSubstringWithAtMostKDistinctCharacters.cpp
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        int high = 0, count = 0, distinct = 0, trail = 0;
        vector<int> letters(126, 0);
        for (int i = 0; i < s.size(); ++i) {
            count++;
            if (++letters[s[i]] == 1) distinct++;
            if (distinct > k) {
                while (trail <= i && distinct > k) {
                    letters[s[trail]]--;
                    if (letters[s[trail++]] == 0) distinct--;
                    count--;
                }
            }
            high = max(count, high);
        }
        return high;
    }
};

class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        if (s.length()*k == 0) return 0;
        
        unordered_map<int, int> let_counts;
        int k_count = 0, trailing_idx = 0, high_count = 1;
        for (int i = 0; i < s.length(); ++i) {
            if (let_counts.find(s[i]) == let_counts.end() || let_counts[s[i]] == 0) {
                ++k_count;
                let_counts[s[i]] = 1;
            } else {
                let_counts[s[i]]++;
            }
            
            while (trailing_idx < s.length() && k_count > k) {
                char first = s[trailing_idx];
                while (trailing_idx < s.length() && first == s[trailing_idx]) ++trailing_idx, --let_counts[first];
                if (let_counts[first] == 0) --k_count;
            }
            if (high_count < i-trailing_idx+1) high_count = i-trailing_idx+1;
        }
        return high_count;
    }
};
